fire/inspectutils.py

```diff
@@ -63,46 +77,40 @@ def GetArgSpec(fn):
   Args:
     fn: The function to analyze.
   Returns:
-    A named tuple of type inspect.ArgSpec with the following fields:
+    An object containing the following fields:
       args: A list of the argument names accepted by the function.
       varargs: The name of the *varargs argument or None if there isn't one.
-      keywords: The name of the **kwargs argument or None if there isn't one.
+      varkw: The name of the **kwargs argument or None if there isn't one.
       defaults: A tuple of the defaults for the arguments that accept defaults.
+      kwonlyargs: A list of argument names that must be passed with a keyword.
+      kwonlydefaults: A dictionary of keyword only arguments and their defaults.
+      annotations: A dictionary of arguments and their annotated types.
   """
   fn, skip_arg = _GetArgSpecFnInfo(fn)
 
   try:
 
     if six.PY2:
       argspec = inspect.getargspec(fn)
-      keywords = argspec.keywords
+      args = argspec.args[1:] if skip_arg else argspec.args
+      annotations = getattr(fn, '__annotations__', None)
+      fullspec = _FullArgSpec(args, *argspec[1:], annotations=annotations)
     else:
       argspec = inspect.getfullargspec(fn)
-      keywords = argspec.varkw
-
-    args = argspec.args
-    defaults = argspec.defaults or ()
-    varargs = argspec.varargs
+      args = argspec.args[1:] if skip_arg else argspec.args
+      fullspec = _FullArgSpec(args, *argspec[1:])
 
   except TypeError:
-    args = []
-    defaults = ()
     # If we can't get the argspec, how do we know if the fn should take args?
     # 1. If it's a builtin, it can take args.
     # 2. If it's an implicit __init__ function (a 'slot wrapper'), take no args.
     # Are there other cases?
```

Comment:
Yeah, callable objects too, like functools.partial. There is a 50/50 chance I am missing something, but I think those are basically all the cases.

You will probably be special casing forever of course. The first one should probably be special logic to grab the real arguments from the partial object, instead of the generic *args, **kwargs it provides.  My assumption is that you will eventually want something that is or looks like single dispatch here.